logsrvd/sendlog.c

```diff
@@ -169,7 +177,125 @@ init_tls_client_context(const char *ca_bundle_file, const char *cert_file, const
 exit:
     return ctx;
 }
-#endif
+
+static void
+tls_connect_cb(int sock, int what, void *v)
+{
+    struct client_closure *closure = v;
+    struct timespec timeo = { TLS_HANDSHAKE_TIMEO_SEC, 0 };
+    int con_stat, err;
+
+    debug_decl(tls_connect_cb, SUDO_DEBUG_UTIL)
+
+    if (what == SUDO_EV_TIMEOUT) {
+        sudo_warnx(U_("TLS handshake timeout occured"));
+        debug_return;
+    }
+
+    con_stat = SSL_connect(ssl);
+
+    if (con_stat == 1) {
+        closure->tls_connect_state = true;
+    } else {
+        switch ((err = SSL_get_error(ssl, con_stat))) {
+            /* TLS connect successful */
+            case SSL_ERROR_NONE:
+                closure->tls_connect_state = true;
+                break;
+            /* TLS handshake is not finished, reschedule event */
+            case SSL_ERROR_WANT_READ:
+            case SSL_ERROR_WANT_WRITE:
+                if (sudo_ev_add(closure->tls_connect_ev->base,
+                    closure->tls_connect_ev, &timeo, false) == -1) {
+                    sudo_warnx(U_("unable to add event to queue"));
+                }
+                break;
+            default:
+                sudo_warnx(U_("SSL_connect failed: ssl_error=%d, stack=%s\n"),
+                    err,
+                    ERR_error_string(ERR_get_error(), NULL));
+                break;
+        }
+    }
+
+    debug_return;
+}
+
+static bool
+tls_connect_async(struct client_closure *closure)
+{
+    struct sudo_event_base *evbase = NULL;
+    struct timespec timeo = { TLS_HANDSHAKE_TIMEO_SEC, 0 };
+
+    debug_decl(tls_connect_async, SUDO_DEBUG_UTIL)
+
+	evbase = sudo_ev_base_alloc();
+	closure->tls_connect_ev = sudo_ev_alloc(closure->sock, SUDO_EV_WRITE,
+        tls_connect_cb, closure);
+    closure->tls_connect_state = false;
+	if (evbase == NULL || closure->tls_connect_ev == NULL) {
+        sudo_warnx(U_("unable to allocate memory"));
+	    goto done;
+	}
+	if (sudo_ev_add(evbase, closure->tls_connect_ev, NULL, false) == -1) {
+	    sudo_warnx(U_("unable to add event to queue"));
+	    goto done;
+	}
+	if (sudo_ev_dispatch(evbase) == -1) {
+	    sudo_warn(U_("error in event loop"));
+	    goto done;
+	}
+
+done:
+    sudo_ev_base_free(evbase);
+    sudo_ev_free(closure->tls_connect_ev);
+
+    debug_return_int(closure->tls_connect_state);
+}
+
+static bool
+do_tls_handshake(struct client_closure *closure)
+{
+    int con_stat, err = -1;
```

Comment:
The con_stat and err variables are unused.

Reply:
i left those accidentally when this part was refactored... fixed!